File: bundle/edu.gemini.ags/src/main/java/edu/gemini/ags/gems/TiptiltFlexurePair.java
package edu.gemini.ags.gems;


import edu.gemini.spModel.gems.GemsGuideProbeGroup;
import edu.gemini.spModel.gems.GemsGuideStarType;

import java.util.ArrayList;
import java.util.List;

/**
 * Groups a pair of GemsCatalogSearchResults to be used for tiptilt and flexure stars.
 * The two results must be from different guide probe groups (Canopus, GSAOI, etc).
 */
public class TiptiltFlexurePair {
    private final GemsCatalogSearchResults tiptiltResults;
    private final GemsCatalogSearchResults flexureResults;

    private TiptiltFlexurePair(GemsCatalogSearchResults tiptiltResults, GemsCatalogSearchResults flexureResults) {
        this.tiptiltResults = tiptiltResults;
        this.flexureResults = flexureResults;
    }

    public GemsCatalogSearchResults getTiptiltResults() {
        return tiptiltResults;
    }

    public GemsCatalogSearchResults getFlexureResults() {
        return flexureResults;
    }

    // Returns pairs of results that can be used for tiptilt and flexure (from different guide probe groups).
    public static List<TiptiltFlexurePair> pairs(final List<GemsCatalogSearchResults> results) {
        GemsCatalogSearchResults canopusTiptilt = null;
        GemsCatalogSearchResults gsaoiOrFlamingosFlexure = null;

        GemsCatalogSearchResults gsaoiTiptilt = null;
        GemsCatalogSearchResults canopusFlexure = null;

        for (GemsCatalogSearchResults searchResults : results) {
            final GemsCatalogSearchCriterion criterion = searchResults.criterion();
            final GemsCatalogSearchKey key = criterion.key();
            if (key.starType() == GemsGuideStarType.tiptilt) {
                final GemsGuideProbeGroup tiptiltGroup = key.group();
                final String groupKey = tiptiltGroup.getKey();
                if ("CWFS".equals(groupKey)) {
                    canopusTiptilt = searchResults;
                } else if ("ODGW".equals(groupKey)) {
                    gsaoiTiptilt = searchResults;
                }
            } else {
                final GemsGuideProbeGroup flexureGroup = key.group();
                final String groupKey = flexureGroup.getKey();
                if ("ODGW".equals(groupKey) || "FII OIWFS".equals(groupKey)) {
                    gsaoiOrFlamingosFlexure = searchResults;
                } else if ("CWFS".equals(groupKey)) {
                    canopusFlexure = searchResults;
                }
            }
        }

        final List<TiptiltFlexurePair> pairs = new ArrayList<>();
        if (canopusTiptilt != null && gsaoiOrFlamingosFlexure != null) {
            pairs.add(new TiptiltFlexurePair(canopusTiptilt, gsaoiOrFlamingosFlexure));
        }
        if (gsaoiTiptilt != null && canopusFlexure != null) {
            pairs.add(new TiptiltFlexurePair(gsaoiTiptilt, canopusFlexure));
        }
        return pairs;
    }
}

File: bundle/edu.gemini.ags/src/main/scala/edu/gemini/ags/gems/TiptiltFlexurePair.scala
package edu.gemini.ags.gems

import edu.gemini.spModel.gems.GemsGuideStarType
import scala.collection.JavaConverters._

import scalaz._
import Scalaz._

case class TiptiltFlexurePair(tiptiltResults: GemsCatalogSearchResults, flexureResults: GemsCatalogSearchResults)

/**
 * Groups a pair of GemsCatalogSearchResults to be used for tiptilt and flexure stars.
 * The two results must be from different guide probe groups (Canopus, GSAOI, etc).
 */
object TiptiltFlexurePair {
  private case class TipTiltFlexure(tt: Option[GemsCatalogSearchResults], flex: Option[GemsCatalogSearchResults])

  def pairs(results: List[GemsCatalogSearchResults]): List[TiptiltFlexurePair] = {
    val pairs = (TipTiltFlexure(None, None), TipTiltFlexure(None, None))
    // Go Over the results and assign them to buckets, it will keep the last result for each subgroup
    val resultPair = results.foldLeft(pairs) { (p, v) =>
      v.criterion.key match {
        case k if k.starType == GemsGuideStarType.tiptilt && k.group.getKey == "CWFS"                                    => (p._1.copy(tt = v.some), p._2)
        case k if k.starType == GemsGuideStarType.tiptilt && k.group.getKey == "ODGW"                                    => (p._1, p._2.copy(tt = v.some))
        case k if k.starType == GemsGuideStarType.flexure && (k.group.getKey == "ODGW" || k.group.getKey == "FII OIWFS") => (p._1.copy(flex = v.some), p._2)
        case k if k.starType == GemsGuideStarType.flexure && k.group.getKey == "CWFS"                                    => (p._1, p._2.copy(flex = v.some))
      }
    }
    // If they have valid results convert them to pairs
    List(resultPair._1, resultPair._2).collect {
      case TipTiltFlexure(Some(tt), Some(flex)) => TiptiltFlexurePair(tt, flex)
    }
  }
}
